File: java/523_Continuous_Subarray_Sum.java
class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        int n = nums.length;
        List<Integer> preSum = new ArrayList<>();
        preSum.add(nums[0]%k);
        Set hashSet = new HashSet<>();

        for(int i = 1; i < n ; i++){
            preSum.add(i,(preSum.get(i-1) + nums[i])%k);
            if (preSum.get(i) % k == 0 || hashSet.contains(preSum.get(i))){
                return true;
            }
            hashSet.add(preSum.get(i-1));
        }
        return false;
    }

    public boolean checkSubarraySum1(int[] nums, int k) {
        int n = nums.length;
        int[] sum = new int[n+1];
        for(int i = 1 ; i <= n; i++){
            sum[i] = (sum[i-1] + nums[i-1])%k;
        }
        Set<Integer> set = new HashSet<>();
        for(int i = 2 ; i <= n ; i++){
            set.add(sum[i-2]);
            if(set.contains(sum[i])){
                return true;
            }
        }
        return false;
    }
}

File: python/523_Continuous_Subarray_Sum.py
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        if n == 0 or k == 0:
            return False
        hashSet = set()
        preSum = [nums[0]%k] + [0] * (n-1)
        for i in range(1,n):
            preSum[i] = ( preSum[i-1] + nums[i] ) % k
            if preSum[i] == 0 or preSum[i] in hashSet:
                return True
            hashSet.add(preSum[i-1])
        return False